src/engine/ContextDependentPiecewiseLinearConstraint.h

```diff
@@ -0,0 +1,246 @@
+ **     segments the constraint has, initialized by the constructor.
+ **  * getAllCases() virtual method returns all possible phases represented using
+ **    PhaseStatus enumeration. The order of returned cases will determine the
+ **    search order, so this method should implement any search heuristics.
+ **/
+
+#ifndef __ContextDependentPiecewiseLinearConstraint_h__
+#define __ContextDependentPiecewiseLinearConstraint_h__
+
+#include "FloatUtils.h"
+#include "ITableau.h"
+#include "List.h"
+#include "Map.h"
+#include "MarabouError.h"
+#include "PiecewiseLinearCaseSplit.h"
+#include "PiecewiseLinearConstraint.h"
+#include "PiecewiseLinearFunctionType.h"
+#include "Queue.h"
+#include "Tightening.h"
+#include "context/cdlist.h"
+#include "context/cdo.h"
+#include "context/context.h"
+
+class Equation;
+class BoundManager;
+class ITableau;
+class InputQuery;
+class String;
+
+class ContextDependentPiecewiseLinearConstraint
+   : public virtual PiecewiseLinearConstraint
+{
+public:
+    ContextDependentPiecewiseLinearConstraint();
+    ContextDependentPiecewiseLinearConstraint( unsigned numCases );
+
+    virtual ~ContextDependentPiecewiseLinearConstraint()
+    {
+        cdoCleanup();
+    }
+
+    /*
+      Turn the constraint on/off.
+    */
+    void setActiveConstraint( bool active );
+    bool isActive() const;
+
+    /*
+       Returns a list of all cases of this constraint and is used by the
+       nextFeasibleCase during search. The order of returned cases affects the
+       search, and this method is where related heuristics should be
+       implemented.
+     */
+    virtual List<PhaseStatus> getAllCases() const = 0;
+
+    /*
+     * Returns case split corresponding to the given phase/id
+       TODO: Update the signature in PiecewiseLinearConstraint, once the new
+       search is integrated.
+     */
+    virtual PiecewiseLinearCaseSplit getCaseSplit( PhaseStatus caseId ) const = 0;
+
+
+    /*
+      Check if the constraint's phase has been fixed.
+    */
+    virtual bool phaseFixed() const = 0;
+
+    /*
+      If the constraint's phase has been fixed, get the (valid) case split.
+    */
+    virtual PiecewiseLinearCaseSplit getImpliedCaseSplit() const = 0;
+
+    /*
+      Register a bound manager. If a bound manager is registered,
+      this piecewise linear constraint will inform the tightener whenever
+      it discovers a tighter (entailed) bound.
+    */
+    void registerBoundManager( BoundManager *boundManager );
+
+    /*
+       Register context object. Necessary for lazy backtracking features - such
+       as _cdPhaseStatus and _activeStatus. Does not require initialization until
+       after pre-processing.
+     */
+    void initializeCDOs( CVC4::context::Context *context );
+
+    /*
+       Politely clean up allocated CDOs.
+     */
+    void cdoCleanup();
+
+    /*
+      Get the context object - debugging purposes only
+    */
+    CVC4::context::Context *getContext() const
```

Comment:
Can we make the returned object const? i.e., const Context *
Same for the other debugging functions that return pointers

Reply:
Fixed.